File: clients/typescript/test/unit/resume.test.ts
import { describe, expect, test } from "vitest";

import { ResumeState } from "../../src/resume.js";

const id = (n: number): string => `00000000-0000-0000-0000-${String(n).padStart(12, "0")}`;

describe("ResumeState", () => {
  test("skips seq at or below the cursor (conformance item 4)", () => {
    const state = new ResumeState(0, 16);
    expect(state.shouldProcess(id(1), 1)).toBe(true);
    state.record(id(1), 1);
    expect(state.cursor()).toBe(1);
    // A replayed duplicate at seq 1 is now below the cursor.
    expect(state.shouldProcess(id(1), 1)).toBe(false);
    // A brand-new lower seq is still guarded.
    expect(state.shouldProcess(id(99), 1)).toBe(false);
    // The next higher seq proceeds.
    expect(state.shouldProcess(id(2), 2)).toBe(true);
  });

  test("dedups by id across the replay/live boundary (conformance item 4)", () => {
    const state = new ResumeState(5, 16);
    expect(state.shouldProcess(id(7), 6)).toBe(true);
    state.record(id(7), 6);
    // Same id delivered twice (once replay, once live): processed once.
    expect(state.shouldProcess(id(7), 6)).toBe(false);
    // Even at a higher seq label, a seen id is skipped.
    expect(state.shouldProcess(id(7), 8)).toBe(false);
  });

  test("cursor advances only via record (conformance item 5)", () => {
    const state = new ResumeState(10, 16);
    expect(state.shouldProcess(id(1), 11)).toBe(true);
    expect(state.cursor()).toBe(10); // asking does not move it
    state.record(id(1), 11);
    expect(state.cursor()).toBe(11);
  });

  test("bounded seen set evicts oldest but seq guard still holds", () => {
    const state = new ResumeState(0, 2);
    state.record(id(1), 1);
    state.record(id(2), 2);
    state.record(id(3), 3); // evicts id(1) from the recent set
    // id(1) evicted from the id set, but seq 1 is below the cursor (3).
    expect(state.shouldProcess(id(1), 1)).toBe(false);
    // id(2) still remembered by id.
    expect(state.shouldProcess(id(2), 2)).toBe(false);
  });

  test("global seq gaps are not treated as loss (item 5 note)", () => {
    // A connection may observe seq 3 then 9 (4..8 went to other patterns).
    const state = new ResumeState(0, 16);
    state.record(id(3), 3);
    expect(state.cursor()).toBe(3);
    expect(state.shouldProcess(id(9), 9)).toBe(true);
    state.record(id(9), 9);
    expect(state.cursor()).toBe(9);
  });
});

File: clients/go/resume_test.go
package whdrsub

import (
	"fmt"
	"testing"
)

func id(n int) string { return fmt.Sprintf("id-%d", n) }

// Conformance item 4: ignore frames with seq <= cursor.
func TestSkipsSeqAtOrBelowCursor(t *testing.T) {
	s := newResumeState(0, 16)
	if !s.shouldProcess(id(1), 1) {
		t.Fatal("seq 1 should process from cursor 0")
	}
	s.record(id(1), 1)
	if s.Cursor() != 1 {
		t.Fatalf("cursor = %d, want 1", s.Cursor())
	}
	// A replayed duplicate at seq 1 is now at/below the cursor.
	if s.shouldProcess(id(1), 1) {
		t.Fatal("duplicate seq 1 should be skipped")
	}
	// A brand-new lower seq is also guarded.
	if s.shouldProcess(id(99), 1) {
		t.Fatal("seq 1 below cursor should be skipped regardless of id")
	}
	if !s.shouldProcess(id(2), 2) {
		t.Fatal("seq 2 above cursor should process")
	}
}

// Conformance item 4: dedup by id across the replay/live boundary.
func TestDedupsByIDAcrossReplayLiveBoundary(t *testing.T) {
	s := newResumeState(5, 16)
	if !s.shouldProcess(id(7), 6) {
		t.Fatal("new event should process")
	}
	s.record(id(7), 6)
	if s.shouldProcess(id(7), 6) {
		t.Fatal("same id at same seq should dedup")
	}
	// id in seen set is skipped even at a higher seq label.
	if s.shouldProcess(id(7), 8) {
		t.Fatal("same id at higher seq should still dedup")
	}
}

// Conformance item 5: the cursor advances only after record (i.e. after handle).
func TestCursorAdvancesOnlyViaRecord(t *testing.T) {
	s := newResumeState(10, 16)
	if !s.shouldProcess(id(1), 11) {
		t.Fatal("seq 11 should process")
	}
	if s.Cursor() != 10 {
		t.Fatalf("cursor moved on shouldProcess: %d", s.Cursor())
	}
	s.record(id(1), 11)
	if s.Cursor() != 11 {
		t.Fatalf("cursor = %d, want 11", s.Cursor())
	}
}

func TestBoundedSeenSetEvictsOldest(t *testing.T) {
	s := newResumeState(0, 2)
	s.record(id(1), 1)
	s.record(id(2), 2)
	s.record(id(3), 3) // evicts id(1)
	// id(1) is evicted from the recent-id set, but its seq (1) is below the
	// cursor (now 3), so it is still guarded.
	if s.shouldProcess(id(1), 1) {
		t.Fatal("evicted id below cursor should still be guarded by seq")
	}
	if s.shouldProcess(id(2), 2) {
		t.Fatal("id(2) still remembered by id")
	}
}

func TestCapacityClampedToOne(t *testing.T) {
	s := newResumeState(0, 0)
	if s.capacity != 1 {
		t.Fatalf("capacity = %d, want clamped to 1", s.capacity)
	}
}
